jax/scipy/optimize/line_search.py

```diff
@@ -0,0 +1,453 @@
+    star_to_j = wolfe_two(dphi_j) & (~hi_to_j)
+    hi_to_lo = (dphi_j * (state.a_hi - state.a_lo) >= 0.) & (~hi_to_j) & (~star_to_j)
+    lo_to_j = (~hi_to_j) & (~star_to_j)
+
+    state = state._replace(**_binary_replace(hi_to_j,
+                                             state._asdict(),
+                                             dict(a_hi=a_j,
+                                                  phi_hi=phi_j,
+                                                  dphi_hi=dphi_j,
+                                                  a_rec=state.a_hi,
+                                                  phi_rec=state.phi_hi),
+                                             ['a_hi', 'phi_hi', 'dphi_hi', 'a_rec', 'phi_rec']))
+
+    # for termination
+    state = state._replace(done=star_to_j | state.done, **_binary_replace(star_to_j,
+                                                                          state._asdict(),
+                                                                          dict(a_star=a_j,
+                                                                               phi_star=phi_j,
+                                                                               dphi_star=dphi_j,
+                                                                               g_star=g_j),
+                                                                          ['a_star', 'phi_star', 'dphi_star',
+                                                                           'g_star']))
+
+    state = state._replace(**_binary_replace(hi_to_lo,
+                                             state._asdict(),
+                                             dict(a_hi=a_lo,
+                                                  phi_hi=phi_lo,
+                                                  dphi_hi=dphi_lo,
+                                                  a_rec=state.a_hi,
+                                                  phi_rec=state.phi_hi),
+                                             ['a_hi', 'phi_hi', 'dphi_hi', 'a_rec', 'phi_rec']))
+
+    state = state._replace(**_binary_replace(lo_to_j,
+                                             state._asdict(),
+                                             dict(a_lo=a_j,
+                                                  phi_lo=phi_j,
+                                                  dphi_lo=dphi_j,
+                                                  a_rec=state.a_lo,
+                                                  phi_rec=state.phi_lo),
+                                             ['a_lo', 'phi_lo', 'dphi_lo', 'a_rec', 'phi_rec']))
+
+    state = state._replace(j=state.j + 1)
+    return state
+
+  state = while_loop(lambda state: (~state.done) & (~pass_through) & (~state.failed),
+                     body,
+                     state)
+
+  return state
+
+
+def line_search(value_and_gradient, position, direction, f_0=None, g_0=None, max_iterations=50, c1=1e-4,
+                c2=0.9):
+  """
```

Comment:
This interface doesn't match that of scipy.optimize.line_search . For example  f_0 is old_fval and g_0 is gfk in the [scipy implementation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.line_search.html).

Reply:
It is usually important to have a maxiter argument to avoid the line-search method from hanging or choosing an abnormally small step-size when some of the quantities are close to machine precision (scipy's one has maxiter=10).

Reply:
Re: first point. I considered this in choosing the names. I find the choice of variable names in scipy's interface non-informative for our use case. For example, the use of the name `old_fval` makes sense to scipy because they use `old_fval` and `old_old_fval` to compute an initial step size estimate, however in our implementation we do not do that so the name makes little sense. Then to keep things homogeneous I chose `f_0` and `g_0`. I could use `fk` and `gfk` if that pleases you. Also, I've changed `max_iterations` to `maxiter`.

Re: second point. The max iterations is not really needed to avoid getting close to zero, since the line search will terminate with a failure to converge flag when the step size gets too small before using up all iterations. I can set it 10 though by default, which will allow a maximum step size of 2^10.

Reply:
1. I agree that scipy's naming is not great but IMO API compatibility is more important

Reply:
I agree with @fabianp that we should use SciPy's names unless we want to refer to a totally different variables (or maybe if the scipy name is really terrible).

Reply:
Ok, done.